simple/manifest_test.go

```diff
@@ -0,0 +1,210 @@
+func randomAddress() string {
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		panic(err)
+	}
+	return hex.EncodeToString(b)
+}
+
+func TestNilPath(t *testing.T) {
+	m := simple.NewManifest()
+	n, err := m.Lookup("")
+	if err == nil {
+		t.Fatalf("expected error, got reference %s", n.Reference())
+	}
+}
+
+// struct for manifest entries for test cases
+type e struct {
+	path      string
+	reference string
+}
+
+var testCases = []struct {
+	name    string
+	entries []e // entries to add to manifest
+}{
+	{
+		name:    "empty-manifest",
+		entries: nil,
+	},
+	{
+		name: "one-entry",
+		entries: []e{
+			{
+				path:      "entry-1",
+				reference: randomAddress(),
+			},
+		},
+	},
+	{
+		name: "two-entries",
+		entries: []e{
+			{
+				path:      "entry-1.txt",
+				reference: randomAddress(),
+			},
+			{
+				path:      "entry-2.png",
+				reference: randomAddress(),
+			},
+		},
+	},
+	{
+		name: "nested-entries",
+		entries: []e{
+			{
+				path:      "text/robots.txt",
+				reference: randomAddress(),
+			},
+			{
+				path:      "img/1.png",
+				reference: randomAddress(),
+			},
+			{
+				path:      "img/2.jpg",
+				reference: randomAddress(),
+			},
+			{
+				path:      "readme.md",
+				reference: randomAddress(),
+			},
+		},
+	},
+}
+
+func TestEntries(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			m := simple.NewManifest()
+
+			checkLength(t, m, 0)
+
+			// add entries
+			for i, e := range tc.entries {
+				err := m.Add(e.path, e.reference)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				checkLength(t, m, i+1)
+				checkEntry(t, m, e.reference, e.path)
+			}
+
+			manifestLen := m.Length()
+
+			if len(tc.entries) != manifestLen {
+				t.Fatalf("expected %d entries, found %d", len(tc.entries), manifestLen)
+			}
+
+			if manifestLen == 0 {
+				// special case for empty manifest
+				return
+			}
+
+			// replace entry
+			lastEntry := tc.entries[len(tc.entries)-1]
+
+			newReference := randomAddress()
+
+			err := m.Add(lastEntry.path, newReference)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			checkLength(t, m, manifestLen) // length should not have changed
+			checkEntry(t, m, newReference, lastEntry.path)
+
+			// remove entries
+			err = m.Remove("invalid/path.ext") // try removing inexistent entry
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			checkLength(t, m, manifestLen) // length should not have changed
+
+			for i, e := range tc.entries {
+				err = m.Remove(e.path)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				entry, err := m.Lookup(e.path)
+				if entry != nil || !errors.Is(err, simple.ErrNotFound) {
+					t.Fatalf("expected path %v not to be present in the manifest, but it was found", e.path)
+				}
+
+				checkLength(t, m, manifestLen-i-1)
+			}
+
+		})
+	}
+}
+
+// checkLength verifies that the given manifest length and integer match.
+func checkLength(t *testing.T, m simple.Manifest, length int) {
```

Comment:
`t.Helper()`?

Reply:
This test code was from before, I will fix it.

Reply:
Done.